### PR TITLE
REF: define related filetypes once.

### PR DESCRIPTION
**Pull request by ashgillman**
In response to #1612

Refactors using DRY principal, define related files once.

Additionally, can be patched to extend the related filtypes with:

    import nipype
    nipype.utils.filemanip.related_filetype_sets.extend([
        ('.avh', '.hv', '.v'),
        ('.avs', '.hs', '.s'),
    ])

Local tests run, and I have checked the above patching with some of my code using Interfiles.

Potential caveats:
 - copyfile is recursive, my base condition is not to recurse if the destination file exists (i.e., has been copied). This could lead to infinite recursion if the copy fails. Suggestions are welcome (check for infinite recursion and throw instructive error? Add a new function argument defining whether to check related files so is only checked at top level?)

- behaviour is not identical to before. i.e., before `copyfile` on checked for related files for `.img` or `.BRIK`, and not if the other file part was used. I think this way is better, but some issues may come from using `.mat` files?

### CODE REVIEW

File: nipype/utils/filemanip.py [hunk 1]
```diff
@@ -29,6 +29,12 @@
 fmlogger = logging.getLogger("filemanip")
 
 
+related_filetype_sets = [
```
Comment (shoshber) [inline nipype/utils/filemanip.py]: Even DRYer--assuming all extensions start with a period, no need to specify the period here!

File: nipype/utils/filemanip.py [hunk 2]
```diff
@@ -314,38 +320,30 @@ def copyfile(originalfile, newfile, copy=False, create_new=False,
...
-            related_files.append(os.path.join(path, name + ext))
-    elif filename.endswith(".BRIK") or filename.endswith(".HEAD"):
-        path, name, ext = split_filename(filename)
-        for ext in ['.BRIK', '.HEAD']:
-            related_files.append(os.path.join(path, name + ext))
+    path, name, ext = split_filename(filename)
+    for type_set in related_filetype_sets:
+        if ext in type_set:
+            for new_ext in type_set:
+                related_files.append(os.path.join(path, name + new_ext))
```
Comment (shoshber) [inline nipype/utils/filemanip.py]: lines 342-345 and 323-326 are the same--consider refactoring into a `get_type_set()` or `has_type_set()` function of some sort

File: nipype/utils/filemanip.py [hunk 3]
```diff
@@ -314,38 +320,30 @@ def copyfile(originalfile, newfile, copy=False, create_new=False,
...
-        hdrnfile = newfile[:-5] + ".HEAD"
-        copyfile(hdrofile, hdrnfile, copy, hashmethod=hashmethod,
-                 use_hardlink=use_hardlink)
+    _, _, this_type = split_filename(originalfile)
+    for type_set in related_filetype_sets:
+        if this_type in type_set:
+            for alt_type in type_set:
+                alt_ofile = originalfile[:-len(this_type)] + alt_type
+                alt_nfile = newfile[:-len(this_type)] + alt_type
+                if os.path.exists(alt_ofile) and not os.path.exists(alt_nfile):
```
Comment (shoshber) [inline nipype/utils/filemanip.py]: this behavior feels a little weird to me. I don't have neuro background, so perhaps it isn't weird at all. But if `alt_ofile` AND `alt_nfile` exist, what does that mean? Should we throw a warning? Should we copy over `alt_nfile`? Can we perhaps use `get_related_files` in an intelligent way here?

**Comment by shoshber** [comment]
I would love to see some unittests here

File: nipype/utils/filemanip.py (hunk 3 above)
Comment (ashgillman) [inline nipype/utils/filemanip.py]: if original and new files exists, the function assumes we have already done the copy. This is therefore, the recursive base condition. This was my attempt at minimising changes to the function signature. Probably a better way is to change the signature:

    def copyfile(originalfile, newfile, copy=False, create_new=False,
                 hashmethod=None, use_hardlink=False, copy_related_files=True)

And then, when we recurse for related files, set `copy_related_files` to `False`

File: nipype/utils/filemanip.py (hunk 1 above)
Comment (ashgillman) [inline nipype/utils/filemanip.py]: I take your point, but I think that leaving the periods makes the intent more obvious.

Would you suggest something like:

    related_filetype_sets = [tuple('.' + type_ext)
                             for type_ext in (
                                 ('hdr', 'img', 'mat'),
                                 ('BRIK', 'HEAD'),
                             )]

Or just handling it later (i.e., in a function `get_related_filetypes` that does what you suggest below)

I am still personally opting for leaving the dots in.

**Comment by ashgillman** [comment]
Good suggestion on using `get_related_filetypes` from `copyfile`.

`copyfile` seems to be tested quite well (except only using analyze files), and I have added tests for `get_related_files`

File: nipype/utils/filemanip.py (hunk 3 above)
Comment (shoshber) [inline nipype/utils/filemanip.py]: ah, I get it. Thanks for the explanation

File: nipype/utils/filemanip.py (hunk 1 above)
Comment (shoshber) [inline nipype/utils/filemanip.py]: I was imagining handling it later--but I'm also not so much of a stickler as to insist on changing this

**Comment by shoshber** [comment]
Good point on the infinite recursion--of your suggestions above, I think I prefer the add-extra-parameters option.

**Comment by shoshber** [comment]
Argh, tiny phone big fingers!

**Comment by ashgillman** [comment]
@shoshber Happens to the best of us. 4ab5554 includes the extra parameter option.

**Comment by shoshber** [comment]
I see that! Looks good to me!

**Comment by oesteban** [comment]
@effigies It looks like we have a comeback of the copy tests failure here (https://s3.amazonaws.com/archive.travis-ci.org/jobs/159763024/log.txt)

```
======================================================================
FAIL: nipype.utils.tests.test_filemanip.test_recopy(os.stat_result(st_mode=33204, st_ino=3540, st_dev=2049, st_nlink=1, st_uid=1000, st_gid=1000, st_size=0, st_atime=1473817459, st_mtime=1473817459, st_ctime=1473817459), os.stat_result(st_mode=33204, st_ino=3540, st_dev=2049, st_nlink=1, st_uid=1000, st_gid=1000, st_size=0, st_atime=1473817460, st_mtime=1473817459, st_ctime=1473817459), 'Symlink - OS: posix; Copy: True; Hardlink: False')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/lib/python3.5/site-packages/numpy/testing/utils.py", line 315, in assert_equal
    assert_equal(actual[k], desired[k], 'item=%r\n%s' % (k, err_msg), verbose)
  File "/home/travis/miniconda/lib/python3.5/site-packages/numpy/testing/utils.py", line 379, in assert_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal:
item=7
Symlink - OS: posix; Copy: True; Hardlink: False
 ACTUAL: 1473817459
 DESIRED: 1473817460

======================================================================
FAIL: nipype.utils.tests.test_filemanip.test_recopy(os.stat_result(st_mode=33204, st_ino=3541, st_dev=2049, st_nlink=1, st_uid=1000, st_gid=1000, st_size=0, st_atime=1473817459, st_mtime=1473817459, st_ctime=1473817459), os.stat_result(st_mode=33204, st_ino=3541, st_dev=2049, st_nlink=1, st_uid=1000, st_gid=1000, st_size=0, st_atime=1473817460, st_mtime=1473817459, st_ctime=1473817459), 'Symlink - OS: posix; Copy: True; Hardlink: False')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/lib/python3.5/site-packages/numpy/testing/utils.py", line 315, in assert_equal
    assert_equal(actual[k], desired[k], 'item=%r\n%s' % (k, err_msg), verbose)
  File "/home/travis/miniconda/lib/python3.5/site-packages/numpy/testing/utils.py", line 379, in assert_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal:
item=7
Symlink - OS: posix; Copy: True; Hardlink: False
 ACTUAL: 1473817459
 DESIRED: 1473817460
```

**Comment by effigies** [comment]
Oh, I see the problem, now. The `atime` may be getting updated if the test straddles a clock tick, even though the files aren't being changed. Check out 05d6a05 (cherry-pickable). It adds a function to ignore `atime` and to wait for clock ticks to make the test deterministic.

That said, it increases the runtime of the test from <1s to ~13s, so you may want to satisfy yourselves that the logic works and then remove the `_wait_for_tick()` calls.

**Comment by shoshber** [comment]
Instead of increasing test time by 13sec, can we just relax the equality requirement? Perhaps with `assert_almost_equal` and epsilon 1

**Comment by effigies** [comment]
I wouldn't want to do `assert_almost_equal` on the whole thing, because inodes that differ by 1 are different inodes. I think removing `atime` from the comparison is the more conservative choice there. And that fully solves the issue here.

The clock-tick waits are basically the proof that the problem is solved, and the debugging technique I used to figure out the problem. The only advantage to leaving them in is that they will alert you if this problem crops up again. I'm perfectly comfortable removing them, if that's an unacceptable price.

**Comment by shoshber** [comment]
Ah, I see. I vote to remove `_wait_for_tick` calls, like you suggest.

**Comment by effigies** [comment]
Cool. Would you rather l rather I submit a separate PR to nipype, a PR to this branch, or just cherry-pick?

The commits to grab: 05d6a05 9573168

**Comment by shoshber** [comment]
I guess a separate PR? Whatever you like best. I tried making it myself and cherrypicking but couldn't figure it out. I may not have the right permissions.

**Pull request merged**